### PR TITLE
Zuul Route Hystrix Fallback

### LINKED ISSUE #250: Zuul - specify fallback per route
Since it is possible to specify Hystrix and Ribbon parameters for each route in Zuul, it would be nice to specify the Hystrix fallback method per route also.

The solution  I have in mind right now:
- programmatic configuration of Hystrix by extending RibbonCommand, where ```getFallback()``` would be overridden
- application.yml: specify the custom class in the ```zuul.<<route>>.hystrix``` configuration
- use the default RibbonCommand if nothing configured

What do you think?

### PR DESCRIPTION
**Pull request by ryanjbaxter**
Fixes #250

### CODE REVIEW

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/route/support/AbstractRibbonCommandFactory.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+public abstract class AbstractRibbonCommandFactory implements RibbonCommandFactory {
+
+	private final Set<ZuulFallbackProvider> fallbackProviders;
+
+	private Map<String, ZuulFallbackProvider> fallbackProviderCache;
+
+	public AbstractRibbonCommandFactory(Set<ZuulFallbackProvider> fallbackProviders){
+		this.fallbackProviders = fallbackProviders;
+		this.fallbackProviderCache = new HashMap<String, ZuulFallbackProvider>();
+	}
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/route/support/AbstractRibbonCommandFactory.java]: Why not eagerly create the cache here?

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/route/apache/HttpClientRibbonCommandFactory.java [hunk 2]
```diff
@@ -16,33 +16,40 @@
 
 package org.springframework.cloud.netflix.zuul.filters.route.apache;
 
+import java.util.Set;
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/route/apache/HttpClientRibbonCommandFactory.java]: should have another newline.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/route/support/AbstractRibbonCommandFactory.java (hunk 1 above)
Comment (ryanjbaxter) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/route/support/AbstractRibbonCommandFactory.java]: You mean populate the cache so its just a lookup when `getFallbackProvider` is called?

Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/route/support/AbstractRibbonCommandFactory.java]: yes

**Comment by ryanjbaxter** [comment]
@spencergibb made the changes you suggested.

When I first started making the changes it felt simple and it could go in 1.2.2.  Looking at what I ended up doing it seemed like a lot.  However, I don't think any of the changes effect backward compatibility so I dont see why we couldnt include in 1.2.2.

**Comment by spencergibb** [comment]
What do our versions mean then?

**Pull request merged**